from controller import Robot, Keyboard
import math

class OttoNinjaController:
    TIME_STEP = 32  # æ¨¡æ“¬æ™‚é–“æ­¥é•·ï¼ŒWebots 'WorldInfo.basicTimeStep' å»ºè­°ä¹Ÿè¨­ç‚º 32ã€‚
    DRIVE_SPEED = 4.0  # è¼ªå­æ¨é€²/æ—‹è½‰é€Ÿåº¦ã€‚

    # è®“è¼ªå­ã€Œç«‹èµ·ä¾†ã€çš„è§’åº¦ï¼ˆåº¦ï¼‰ã€‚é€™å€‹å€¼è‡³é—œé‡è¦ï¼Œè«‹ç²¾ç¢ºèª¿æ•´ï¼
    # é€šå¸¸ -90 åº¦æœƒè®“è…³åº•å‚ç›´æ–¼åœ°é¢ã€‚å¦‚æœæ©Ÿå™¨äººä¸ç©©ï¼Œè«‹å¾®èª¿æ­¤å€¼ã€‚
    STAND_ANGLE_DEG = -90 
    
    # Y è»¸é¦¬é”æ—‹è½‰é€Ÿåº¦ï¼Œå½±éŸ¿ç«™ç«‹å’Œæ¢å¾©å‹•ä½œçš„é€Ÿåº¦èˆ‡å¹³ç©©æ€§ã€‚
    # å¢åŠ æ­¤å€¼å¯ä½¿å‹•ä½œæ›´å¿«ï¼Œä½†è‹¥å¤ªé«˜æœƒå°è‡´ä¸ç©©å’Œè­¦å‘Šã€‚
    STAND_MOTOR_VELOCITY = 0.8 # å¾ 0.3 å¢åŠ åˆ° 0.8ã€‚å¦‚æœä»ç„¶ä¸ç©©ï¼Œè«‹é™ä½æ­¤å€¼ã€‚

    def __init__(self):
        self.robot = Robot()
        self.keyboard = self.robot.getKeyboard()
        self.keyboard.enable(self.TIME_STEP)

        self.right_motory = self.get_dev('right_motory')
        self.left_motory = self.get_dev('left_motory')
        self.right_motorz = self.get_dev('right_motorz')
        self.left_motorz = self.get_dev('left_motorz')

        # è¨­å®š Y è»¸é¦¬é”çš„é€Ÿåº¦ï¼Œç”¨æ–¼ç«™ç«‹æˆ–æ¢å¾©å¹³èººã€‚
        self.right_motory.setVelocity(self.STAND_MOTOR_VELOCITY)
        self.left_motory.setVelocity(self.STAND_MOTOR_VELOCITY)
        
        # åˆå§‹ç‹€æ…‹ï¼šè¼ªå­å¹³èººã€‚
        self.set_leg_y_angle(0, 0)
        
        # è¨­å®š Z è»¸é¦¬é”ç‚ºç„¡é™ä½ç½®æ¨¡å¼ (é€Ÿåº¦æ§åˆ¶)ï¼Œåˆå§‹é€Ÿåº¦ç‚º 0ã€‚
        for wheel in (self.right_motorz, self.left_motorz):
            wheel.setPosition(float('inf'))
            wheel.setVelocity(0)

        self.is_standing = False 
        print("ğŸ¤– Otto-Ninja æ§åˆ¶å™¨å•Ÿå‹•ã€‚")
        print("æŒ‰ä¸‹ 'S' éµè®“æ©Ÿå™¨äººç«™ç«‹ã€‚")
        print("ç«™ç«‹å¾Œï¼Œä½¿ç”¨æ–¹å‘éµæ§åˆ¶ï¼šä¸Š/ä¸‹å‰é€²å¾Œé€€ï¼Œå·¦/å³è½‰å½ã€‚")
        print("æŒ‰ä¸‹ 'Space' éµåœæ­¢ï¼ŒæŒ‰ä¸‹ 'R' éµæ¢å¾©å¹³èººã€‚")

    def get_dev(self, name):
        """ç²å– Webots è£ç½®ã€‚å¦‚æœæ‰¾ä¸åˆ°è£ç½®ï¼Œå‰‡æ‹‹å‡º RuntimeErrorã€‚"""
        d = self.robot.getDevice(name)
        if d is None:
            raise RuntimeError(f'ç„¡æ³•æ‰¾åˆ°è£ç½®: {name}')
        return d

    def set_leg_y_angle(self, right_deg, left_deg):
        """è¨­å®šå·¦å³è…¿éƒ¨ Y è»¸çš„è§’åº¦ã€‚è§’åº¦ä»¥åº¦ç‚ºå–®ä½ï¼Œæœƒè½‰æ›ç‚ºå¼§åº¦ã€‚"""
        self.right_motory.setPosition(math.radians(right_deg))
        self.left_motory.setPosition(math.radians(left_deg))

    def set_z_speeds(self, left_speed, right_speed):
        """è¨­å®šå·¦å³è…³ Z è»¸é¦¬é”çš„é€Ÿåº¦ã€‚"""
        self.left_motorz.setVelocity(left_speed)
        self.right_motorz.setVelocity(right_speed)

    def run(self):
        """æ©Ÿå™¨äººçš„ä¸»æ§åˆ¶è¿´åœˆã€‚"""
        while self.robot.step(self.TIME_STEP) != -1:
            key = self.keyboard.getKey()

            if not self.is_standing:
                # æ©Ÿå™¨äººå°šæœªç«™ç«‹
                if key == ord('S'):
                    print("ğŸš€ Otto-Ninja ç«™ç«‹ä¸­...")
                    # æ ¹æ“šæ‚¨ç¢ºèªçš„ Y è»¸é¦¬é”æ–¹å‘ï¼Œé€™æ¨£è¨­ç½®å¯ä»¥è®“è¼ªå­æ­£ç¢ºç«‹èµ·ä¾†
                    self.set_leg_y_angle(self.STAND_ANGLE_DEG, -self.STAND_ANGLE_DEG)
                    self.is_standing = True
                    
                    # æ¸›å°‘ç«™ç«‹å¾Œçš„ç­‰å¾…æ™‚é–“ï¼Œå¦‚æœæ©Ÿå™¨äººç«™ç«‹ç©©å®šï¼Œå¯ä»¥å°‡å…¶è¨­ç‚º0
                    # å¦‚æœä»æœ‰è­¦å‘Šï¼Œå‰‡éœ€è¦å¢åŠ æ­¤å€¼æˆ–é™ä½ STAND_MOTOR_VELOCITY
                    for _ in range(50): # å¾ 150 æ­¥æ¸›å°‘åˆ° 50 æ­¥ (ç´„ 1.6 ç§’)
                        if self.robot.step(self.TIME_STEP) == -1: return 
                    print("ğŸ¤– Otto-Ninja å·²ç«™ç«‹ï¼Œå¯ä»¥ä½¿ç”¨æ–¹å‘éµæ§åˆ¶ã€‚")
                else:
                    self.set_z_speeds(0, 0)
            else:
                # æ©Ÿå™¨äººå·²ç«™ç«‹ï¼Œæ ¹æ“šæ–¹å‘éµæ§åˆ¶ Z è»¸é¦¬é”
                current_left_speed = 0
                current_right_speed = 0

                if key == Keyboard.UP:
                    current_left_speed = self.DRIVE_SPEED
                    current_right_speed = self.DRIVE_SPEED
                elif key == Keyboard.DOWN:
                    current_left_speed = -self.DRIVE_SPEED
                    current_right_speed = -self.DRIVE_SPEED
                elif key == Keyboard.LEFT:
                    current_left_speed = -self.DRIVE_SPEED
                    current_right_speed = self.DRIVE_SPEED
                elif key == Keyboard.RIGHT:
                    current_left_speed = self.DRIVE_SPEED
                    current_right_speed = -self.DRIVE_SPEED
                elif key == ord(' '): # æŒ‰ä¸‹ç©ºç™½éµåœæ­¢
                    print("ğŸ›‘ åœæ­¢ã€‚")
                    current_left_speed = 0
                    current_right_speed = 0
                elif key == ord('R'): # æŒ‰ä¸‹ 'R' éµæ¢å¾©åˆå§‹å¹³èººç‹€æ…‹
                    print("ğŸ›‹ï¸ æ¢å¾©å¹³èººç‹€æ…‹ã€‚")
                    self.set_leg_y_angle(0, 0) # è®“ Y è»¸é¦¬é”å›åˆ°å¹³èººè§’åº¦
                    self.set_z_speeds(0, 0)    # Z è»¸é¦¬é”åœæ­¢
                    self.is_standing = False
                    
                    # æ¸›å°‘æ¢å¾©å¹³èººå¾Œç­‰å¾…æ™‚é–“ï¼ŒåŒä¸Š
                    for _ in range(50): # å¾ 100 æ­¥æ¸›å°‘åˆ° 50 æ­¥
                        if self.robot.step(self.TIME_STEP) == -1: return
                
                self.set_z_speeds(current_left_speed, current_right_speed)

# åŸ·è¡Œæ§åˆ¶å™¨
if __name__ == "__main__":
    controller = OttoNinjaController()
    controller.run()